### DISCUSSION TITLE
How to extract data as dict from tree widget?

### DISCUSSION THREAD

**Discussion by YoonSungHyun-Git** [Textualize/textual#5224] [Q&A] [answered]
Hello,

It is very simple to create a tree by inserting dict data into the tree widget. 
Conversely, I am curious about how to extract data from a tree into a dict.

**Comment by TomJGooding** [comment]
Sorry, I'm not exactly sure what you mean by "extract data as dict from tree widget". Could you provide a simple example?

**Reply by YoonSungHyun-Git** [reply]
Sorry. I had forgotten that it had a non-basic tree structure.
What I wanted to do was put yml into the tree, modify the values, receive the data from the tree as a dict, and save it as a file.

As an example, the data is as follows:

```
{
  'version' : 1.0.0,
  'title' : 'test'
}
```

If you put this in the tree, it will come out like this:

```
root title
  L version=1.0.0
  L title='test'
```

Adjust the label of the tree so that the key and value are visible based on = and open an input window to edit when clicked.
Afterwards, a recursive call was made to extract the data and the dict was extracted.
I left an inquiry to see if there was a better way.

**Reply by YoonSungHyun-Git** [reply]
Can I ask another question?
You can get the value of the tree by using the tree's label. 
Is there a way to receive the labels of the entire tree as a dict or list?

**Reply by TomJGooding** [reply, marked as answer]
Perhaps rather than traversing the entire tree, you could utilise the [`TreeNode.data`](https://textual.textualize.io/widgets/tree/#textual.widgets.tree.TreeNode)?

Based on your example, you could create the tree something like this. 

```python
for key, value in example_data.items():
    tree.root.add_leaf(
	label=f"{key}={value}",
	data={key: value},
```

Then after the node value has been updated:

```python
example_data.update(node.data)
```

**Reply by YoonSungHyun-Git** [reply]
Development has been completed to add nodes using stack. thank you

```python
@classmethod
    def add_json(cls, node: TreeNode, json_data: object) -> None:
        """Adds JSON data to a node."""
        from rich.highlighter import ReprHighlighter

        highlighter = ReprHighlighter()
        
        stack = [(node, json_data)]

        while stack:
            current_node, current_data = stack.pop()

            if isinstance(current_data, dict):
                for key, value in current_data.items():
                    new_node = current_node.add("")
                    new_node.set_label(Text(f"{key}"))
                    stack.append((new_node, value))
            elif isinstance(current_data, list):
                for index, value in enumerate(current_data):
                    new_node = current_node.add("")
                    new_node.set_label(Text(f"[{index}]"))
                    stack.append((new_node, value))
            else:
                key = str(current_node.label)

                if key:
                    current_node.set_label(Text.from_markup(f"[b]{key}[/b]=") + highlighter(repr(current_data)))
                else:
                    current_node.set_label(Text(repr(current_data)))

                current_node.allow_expand = False
```
